viewing_party/party.py

```diff
@@ -1,23 +1,151 @@
 # ------------- WAVE 1 --------------------
 
+# I feel like there's a better way to test if title, genre, & rating are truthy 
+# but not sure how
+
 def create_movie(title, genre, rating):
-    pass
+    new_movie = {"title": title, "genre": genre, "rating": rating}
+    if title and genre and rating:
+        return new_movie
+    else:
+        return None
+
+
+def add_to_watched(user_data, movie):
+    has_watched = [movie] 
+    user_data = {"watched" : has_watched }
+    return user_data
+
 
+def add_to_watchlist(user_data, movie):
+    wants_to_watch = [] 
+    user_data = {"watchlist" : wants_to_watch }
+    wants_to_watch.append(movie)
+
+    return user_data
+
+def watch_movie(user_data, title):
+    for movies in user_data["watchlist"]:
+        if movies["title"] == title:
+            user_data["watchlist"].remove(movies)
+            user_data["watched"].append(movies)
+    return user_data
+        
+        
 # -----------------------------------------
-# ------------- WAVE 2 --------------------
 # -----------------------------------------
+# ------------- WAVE 2 --------------------
+# -----------------------------------------        
+
+def get_watched_avg_rating(user_data):
+    count = 0
+    movie_rating = 0.0
+    if len(user_data["watched"]) == 0:
+        return 0.0
+    else:
+        for movies in user_data["watched"]:
+            movie_rating += movies["rating"]
+            count += 1  
+        return movie_rating/count
+
+
 
+def get_most_watched_genre(user_data):
+    genre_list = []
+    if len(user_data["watched"]) == 0:
+        return None
+    else:
+        for movie in user_data["watched"]:
+            genre_list.append(movie["genre"])
+        return max(set(genre_list), key = genre_list.count)
 
+        
+
+
+
+# -----------------------------------------
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
+def get_unique_watched(user_data):
+    movies_list = [movies for movies in user_data["watched"]]
+    for movie in user_data["friends"]:
+        for title in movie["watched"]:
+            if title in movies_list:
+                movies_list.remove(title)

```

Comment:
Instead of creating movies_list and removing `title` from it, how could you loop through the movies to check that only the user has watched it and then append that uniquely watched movie to a list instead?

Recall that remove has time complexity of O(n) and you're calling the function inside a for loop already, so you've increased the time complexity. If you loop and check for the unique user movies and append to a list, you can lose that O(n) from remove.

Acutally if we have a look at lines 73 -76, we have several levels of complexity. First we loop through the movies in the friends list, then we loop through the titles in watched list, also recall that `element in list` is also O(n) in the worse case, then `remove` is also O(n). 

When we're working with lists, time complexity has a way of creeping up on us